feat(server): Support webm videos

samip5 opened the pull request:
This PR includes:

- Support webm videos to be uploaded.
- Add a test case for webm



Closes #1248

samip5 committed:
feat(server): Support webm without transcoding.

Transcoding result doesn't appear to be used by anything expect for quicktime.

samip5 committed:
feat(server): Fix the asset uploader for .avi

It needs to be transcoded.

zackpollard commented on server/apps/microservices/src/processors/asset-uploaded.processor.ts:
```diff
@@ -40,7 +40,9 @@ export class AssetUploadedProcessor {
     // Video Conversion
     if (asset.type == AssetType.VIDEO) {
-      await this.videoConversionQueue.add(JobName.MP4_CONVERSION, { asset });
+      if (asset.mimeType == 'video/quicktime') {
```
In my opinion we shouldn't be making these kind of decisions here. The processing of only quicktime files right now is something we are looking to address, which will be done by the video conversion queue. The video conversion queue should just receive all video files and it can make the decision on whether the file actually needs converting or not.

samip5 replied:
How should it make that decision?

samip5 replied:
Currently the transcoder seems to transcode every file that is not mp4.

jrasm91 replied:
I think we're saying we can still choose to not convert videos, but "skipping" should happen in one place and that place should be the mp4 processor implementation itself.

samip5 replied:
I have removed the change here, but I'm wondering how should we handle iOS as it doesn't support webm direct playback apparently?

samip5 committed:
feat(server): Most browsers doesn't support avi so use mp4.

alextran1502 commented on server/apps/immich/src/api-v1/asset/asset.service.ts:
```diff
@@ -456,7 +456,7 @@ export class AssetService {
         await fs.access(videoPath, constants.R_OK | constants.W_OK);
 
-        if (query.isWeb && asset.mimeType == 'video/quicktime') {
+        if (query.isWeb && asset.mimeType == 'video/quicktime' || query.isWeb && asset.mimeType == "video/x-msvideo") {
```
Is this change related to this PR?

samip5 replied:
Yes, because previously all files went to get transcoded to MP4 and this PR changed it but .avi files need the transcode as well due to browser support not existing.

jrasm91 replied:
I think this is a good catch - there's basically a discrepancy between when we transcode a video and when we use the transcoding. Can you make a new function `isWebPlayable(mimeType: string)`. Maybe we can put it in `asset-upload.config.ts` for now, which returns true for all types besides `video/mp4`? And then use that function here (`asset.service.ts`) as well as in the mp4 processor?

samip5 replied:
I have now moved it to a central location and transcoder uses that too.

alextran1502 commented:
I've run some tests, WebM video can be played on the Web, and Android, but cannot play be played on iOS. You might have to convert WebM to MP4. Let me know if you can troubleshoot this issue

alextran1502 commented:
> Hey, thanks for the contribution. WebM support makes sense to me, did you test out how the application handles exif from these types of files with these changes?

FYI, the sample file information panel

![image](https://user-images.githubusercontent.com/27055614/213756804-6deac8b9-97c8-4e98-8617-b7606c7e06dd.png)

samip5 commented:
> I've run some tests, WebM video can be played on the Web, and Android, but cannot play be played on iOS. You might have to convert WebM to MP4. Let me know if you can troubleshoot this issue

Unfortunately I don't have an iOS device so I'm not able to troubleshoot it on that platform and thus why I haven't tested it at all there. (Probably possible somehow to create a iOS emulator on a Mac though?)

I think transcode might be needed, but the API response doesn't usually use the transcoded file except for QuickTime video, or am I mistaken about that?

samip5 commented:
> Hey, thanks for the contribution. WebM support makes sense to me, did you test out how the application handles exif from these types of files with these changes?

I did test it and as Alex showed it does properly process the exif data for/from them too.

samip5 committed:
feat(server): Address PR comments

samip5 committed:
Addressed the PR comments

I moved the function that checks the mimetype to a central location in asset-utils and made tests for it.

samip5 committed:
Merge remote-tracking branch 'origin' into feat/webm-support

samip5 commented:
I have now tested my changes with the iOS simulator, and found the same as Alex, so how do we want to handle it?
It would also be great if the iOS app identified itself when sending a request to the api that it's in fact the iOS client that's requesting it rather than `Web`.

samip5 committed:
Merge remote-tracking branch 'origin' into feat/webm-support

samip5 committed:
Rollbacked to the way transcoder was decising things to transcode.

samip5 commented:
I believe that this can be merged, as #1370 will better handle the transcoding aspect.
